### PR TITLE
CR-1093962  ASTeR TC 7.15 - u50 nodma - xbutil2 m2m test fails with "…

### PR DESCRIPTION
**Pull request by sadasivanshaiju**
…Couldn't sync BO"

### CODE REVIEW

**Comment by sadasivanshaiju** [comment]
There is  a  bug in xbutil2  that M2M  can  run on Non dma shell .  But  on  xbutil  it is  disabled .  I  have  ported  the  same logic to xbutil2. Hope this clear your  question.

Regards,
Shaiju.

**Review by AShivangi — APPROVED** [review]
This looks good

File: src/runtime_src/core/tools/xbutil2/SubCmdValidate.cpp [hunk 1]
```diff
@@ -964,6 +964,15 @@ m2mTest(const std::shared_ptr<xrt_core::device>& _dev, boost::property_tree::ptr
     return;
   }
 
+
+ int nodma   = xrt_core::device_query<xrt_core::query::nodma>(_dev);
+
+  if (nodma == 1 ) {
+    logger(_ptTest, "Details", "M2M is not available");
```
[the remark this answers is not in the document]
Comment (xdavidz) [reply src/runtime_src/core/tools/xbutil2/SubCmdValidate.cpp:971]: @sadasivanshaiju you can comment on github, so the conversation looks nicer.
I think Lizhi's comment is valide. The M2M driver is enable on nodma shell, but running m2mtest on nodma shell should be disabled. So "M2M test is not available" seems better.

**Review by xdavidz — CHANGES_REQUESTED** [review]
@sadasivanshaiju Please test this fix on u250, u50 and u30 if possible. I don't think pipeline will catch up those tests yet. And you can post test results into the comments.

File: src/runtime_src/core/tools/xbutil2/SubCmdValidate.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (sadasivanshaiju) [reply src/runtime_src/core/tools/xbutil2/SubCmdValidate.cpp]: I  have pushed the  change suggested

**Comment by sadasivanshaiju** [comment]
root@xsjyliu50:~# xbutil --new validate -r "Memory to memory DMA"
-----------------------------------------------------
 Invoking next generation of the xbutil application 
-----------------------------------------------------
Starting validation for 2 devices

Validate device[0000:d9:00.1]
Platform            : xilinx_u50_gen3x16_nodma_base_1
SC Version          : 5.2.6
Platform ID         : 0x0

1/1 Test #1 [0000:d9:00.1]  : Memory to memory DMA 
    Description             : Run M2M test
    Details                 : M2M is not available
    Test Status             : [SKIPPED]
-------------------------------------------------------------------------------
Validation completed
Validate device[0000:b1:00.1]
Platform            : xilinx_vck5000-es1_gen3x16_base_2
SC Version          : 4.4.6
Platform ID         : 0x0

1/1 Test #1 [0000:b1:00.1]  : Memory to memory DMA 
    Description             : Run M2M test
    Details                 : M2M is not available
    Test Status             : [SKIPPED]
-------------------------------------------------------------------------------
Validation completed

**Comment by sadasivanshaiju** [comment]
> @sadasivanshaiju Please test this fix on u250, u50 and u30 if possible. I don't think pipeline will catch up those tests yet. And you can post test results into the comments.

root@xsjyliu50:~# xbutil --new validate -r "Memory to memory DMA"
-----------------------------------------------------
 Invoking next generation of the xbutil application 
-----------------------------------------------------
Starting validation for 2 devices

Validate device[0000:d9:00.1]
Platform            : xilinx_u50_gen3x16_nodma_base_1
SC Version          : 5.2.6
Platform ID         : 0x0

1/1 Test #1 [0000:d9:00.1]  : Memory to memory DMA 
    Description             : Run M2M test
    Details                 : M2M is not available
    Test Status             : [SKIPPED]
-------------------------------------------------------------------------------
Validation completed
Validate device[0000:b1:00.1]
Platform            : xilinx_vck5000-es1_gen3x16_base_2
SC Version          : 4.4.6
Platform ID         : 0x0

1/1 Test #1 [0000:b1:00.1]  : Memory to memory DMA 
    Description             : Run M2M test
    Details                 : M2M is not available
    Test Status             : [SKIPPED]
-------------------------------------------------------------------------------
Validation completed

**Review by AShivangi — CHANGES_REQUESTED** [review]
Looks like something got messed up... Can you please clean your PR, my changes are also showing up in here which shouldn't be the case

**Review by houlz0507 — APPROVED** [review]

**Review by AShivangi — APPROVED** [review]

**Review by rozumx — APPROVED** [review]

**Comment by sadasivanshaiju** [comment]
There is  no regression with u250

root@pranjal-dell:/proj/rdi/staff/sadasiva/XRT/build/Release# xbutil --new validate -r "Memory to memory DMA"
-----------------------------------------------------
 Invoking next generation of the xbutil application 
-----------------------------------------------------
Starting validation for 1 devices

Validate device [0000:17:00.1]
Platform            : xilinx_u250_gen3x16_xdma_shell_4_1
SC Version          : 4.6.11
Platform ID         : 0x0
-------------------------------------------------------------------------------
Test 1 [0000:17:00.1]       : Memory to memory DMA 
    Details                 : bank0 -> bank1 M2M bandwidth: 12316.29 MB/s
                              bank0 -> bank2 M2M bandwidth: 12316.60 MB/s
                              bank0 -> bank3 M2M bandwidth: 12323.48 MB/s
                              bank1 -> bank2 M2M bandwidth: 12322.04 MB/s
                              bank1 -> bank3 M2M bandwidth: 12338.82 MB/s
                              bank2 -> bank3 M2M bandwidth: 12310.84 MB/s
    Test Status             : [PASSED]
-------------------------------------------------------------------------------
Validation completed
Validation Summary
------------------
1  device(s) evaluated
1  device(s) validated successfully
0  device(s) had exceptions during validation

Validated successfully [1 device(s)]
  - [0000:17:00.1] : xilinx_u250_gen3x16_xdma_shell_4_1

Validation Exceptions [0 device(s)]

Warnings produced during test [0 device(s)] (Note: The given test successfully validated)
root@pranjal-dell:/proj/rdi/staff/sadasiva/XRT/build/Release# xbutil --new validate 
-----------------------------------------------------
 Invoking next generation of the xbutil application 
-----------------------------------------------------
Starting validation for 1 devices

Validate device [0000:17:00.1]
Platform            : xilinx_u250_gen3x16_xdma_shell_4_1
SC Version          : 4.6.11
Platform ID         : 0x0
-------------------------------------------------------------------------------
Test 1 [0000:17:00.1]       : Aux connection 
    Test Status             : [PASSED]
-------------------------------------------------------------------------------
Test 2 [0000:17:00.1]       : PCIE link 
    Test Status             : [PASSED]
-------------------------------------------------------------------------------
Test 3 [0000:17:00.1]       : SC version 
    Test Status             : [PASSED]
-------------------------------------------------------------------------------
Test 4 [0000:17:00.1]       : Verify kernel 
[PASSED] : Test duration: < 3s >
    Test Status             : [PASSED]
-------------------------------------------------------------------------------
XRT build version: 2.11.0
Build hash: 8a3406ac25fef1a2f47b22de6e1d74e6eaa71322
Build date: 2021-04-23 14:55:23
Git branch: bug_fix
PID: 5492
UID: 0
[Fri Apr 23 23:23:22 2021 GMT]
HOST: pranjal-dell
EXE: /opt/xilinx/xrt/bin/unwrapped/xbutil2
[xbutil] ERROR: No DMA buffers could be allocated.: Cannot allocate memory

root@pranjal-dell:/proj/rdi/staff/sadasiva/XRT/build/Release# xbutil examine
-----------------------------------------------------
 Invoking next generation of the xbutil application 
-----------------------------------------------------
System Configuration
  OS Name              : Linux
  Release              : 4.15.0-142-generic
  Version              : #146-Ubuntu SMP Tue Apr 13 01:11:19 UTC 2021
  Machine              : x86_64
  CPU Cores            : 16
  Memory               : 15727 MB
  Distribution         : Ubuntu 18.04.5 LTS
  GLIBC                : 2.27
  Model                : Precision 5820 Tower

XRT
  Version              : 2.11.0
  Branch               : bug_fix
  Hash                 : 8a3406ac25fef1a2f47b22de6e1d74e6eaa71322
  Hash Date            : 2021-04-23 14:55:23
  XOCL                 : 2.11.0, 8a3406ac25fef1a2f47b22de6e1d74e6eaa71322
  XCLMGMT              : 2.11.0, 8a3406ac25fef1a2f47b22de6e1d74e6eaa71322

Devices present
  [0000:17:00.1] : xilinx_u250_gen3x16_xdma_shell_4_1

**Comment by xdavidz** [comment]
> @sadasivanshaiju Please test this fix on u250, u50 and u30 if possible. I don't think pipeline will catch up those tests yet. And you can post test results into the comments.

I asked Shaiju to verify u250, it seems there is an existing bug even with XRT master code:
HOST: pranjal-dell
EXE: /opt/xilinx/xrt/bin/unwrapped/xbutil2
[xbutil] ERROR: No DMA buffers could be allocated.: Cannot allocate memory

Since it seems not related to this m2m changes, approve this PR.

**Review by xdavidz — APPROVED** [review]

**Pull request merged**